astropy/cosmology/connect.py

```diff
@@ -0,0 +1,99 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import copy
+import warnings
+
+from astropy.io import registry as io_registry
+from astropy.utils.exceptions import AstropyUserWarning
+
+__all__ = ["CosmologyRead", "CosmologyWrite"]
+__doctest_skip__ = __all__
+
+
+class CosmologyRead(io_registry.UnifiedReadWrite):
+    """Read and parse data to a `~astropy.cosmology.Cosmology`.
+
+    This function provides the Cosmology interface to the Astropy unified I/O
+    layer. This allows easily reading a file in supported data formats using
+    syntax such as::
+
+        >>> from astropy.cosmology import Cosmology
+        >>> cosmo1 = Cosmology.read('[file name]')
+
+    When the ``read`` method is called from a subclass the subclass will
+    provide a keyword argument ``cosmology=[class]`` to the registered read
+    method. The method uses this cosmology class, regardless of the class
+    indicated in the file, and sets parameters' default values from the class'
+    signature.
+
+    Get help on the available readers using the ``help()`` method::
+
+      >>> Cosmology.read.help()  # Get help reading and list supported formats
+      >>> Cosmology.read.help('[format]')  # Get detailed help on a format
+      >>> Cosmology.read.list_formats()  # Print list of available formats
+
+    See also: https://docs.astropy.org/en/stable/io/unified.html
+
+    Parameters
+    ----------
+    *args
+        Positional arguments passed through to data reader. If supplied the
+        first argument is typically the input filename.
+    format : str (optional, keyword-only)
+        File format specifier.
+    **kwargs
+        Keyword arguments passed through to data reader.
+
+    Returns
+    -------
+    out : `~astropy.cosmology.Cosmology` subclass instance
+        `~astropy.cosmology.Cosmology` corresponding to file contents.
+    """
+
+    def __init__(self, instance, cosmo_cls):
+        super().__init__(instance, cosmo_cls, "read")
+
+    def __call__(self, *args, **kwargs):
+        from astropy.cosmology.core import Cosmology
+
+        # so subclasses can override also pass the class as a kwarg.
+        if self._cls is not Cosmology:
+            kwargs["cosmology"] = self._cls
```

Comment:
This is indeed a nice & simple solution!! One question: would this be better as `kwargs.setdefault('cosmology', self._cls)`? With the present scheme, user input of 'cosmology' simply gets overwritten, which seems a bit strange.

Reply:
I was thinking about both options for a while. I'm still not 100% certain, but in the end the following example swayed me:
``FlatLambdaCDM.read(..., cosmology=w0wzCDM)``

The only point to using ``FlatLambdaCDM.read`` is to get the cosmology and defaults from ``FlatLambdaCDM``. If one wants those from ``w0wzCDM``, then ``Cosmology.read(..., cosmology=w0wzCDM)`` or ``w0wzCDM.read`` should be used instead.

Reply:
I wonder if in that case one shouldn't simply raise an error if the user passes in `cosmology=...`?

Reply:
Done! and updated the tests.